### ISSUE TITLE
PDF with different kinds of xref table should be invalid

### ISSUE THREAD

**Issue by calixteman** [qpdf/qpdf#1148]
The following pdf:
[plop.pdf](https://github.com/qpdf/qpdf/files/14237823/plop.pdf)

has a xref stream and has been updated with a usual xref.
The preflight tool from Acrobat is unable to read the file because of damaged structure.
Because of that, a pdf can be incorrectly rendered in either Chrome (https://bugs.chromium.org/p/pdfium/issues/detail?id=2123&q=&can=1) or Acrobat (I can't share the pdf here but there's a bug on the Firefox side: https://bugzilla.mozilla.org/show_bug.cgi?id=1878916).
I didn't find anything in the pdf specs about that, but we fixed a bug in pdf.js to avoid to mix different kinds of tables when updating a pdf:
https://github.com/mozilla/pdf.js/pull/17636

`qpdf --check` doesn't show any issues.

**Comment by jberkenbilt** [comment]
There's no reason that you can't mix and match different types of xref tables, and qpdf explicitly supports that, though with the exception of hybrid-reference files (Section 7.5.8.4), which have redundant information between xref tables and xref streams with potentially more objects in xref streams, I agree that the spec is not clear about this. In any case, the attached file is not a hybrid-xref file, just a file that has a mixture of different types of xref tables.

In this file, the last xref table has an incorrect value for /Prev, and the file appears damaged in other ways as other xref previous pointers are also incorrect. I'm not sure why qpdf isn't complaining about the file. It probably should be, but it's possible that it is using heuristics to find %%EOF markers without warning if they're in the wrong place. I'll take a deeper look at the file when I get a chance to see whether qpdf is recovering it correctly. Loading the file with a handful of readers (atril, ghostscript) shows the same appearance before and after processing with qpdf, which suggests qpdf may be recovering in the same way as these other viewers, but I'm still not sure why it's not giving any warnings.

**Comment by jberkenbilt** [comment]
I take it back. The file doesn't contain errors. I think qpdf is correctly interpreting the file and that there's nothing wrong with it. Tracing the file, it says xref is at 10352. That is an xref table whose /Prev is 91. Although the xref stream doesn't start until 116, there is only white space between 91 and 116. This is the first section of a linearized file. Its /Prev points forward to 4618. That is another xref stream followed by a startxref that points to 116, but we never need to look there since this is no longer the last %%EOF.

So, bottom line, the file is fine, and I don't think there's anything to fix here.

**Comment by jberkenbilt** [comment]
I'm leaving the issue open in case there's some other thing going on that is not reproduced with the file you attached.

**Comment by jberkenbilt** [comment]
qpdf doesn't have any issue with the file from the chrome bug.

**Comment by jberkenbilt** [comment]
I would consider a possible enhancement to qpdf to issue a warning if a file mixes and matches in a situation other than hybrid-ref, but I'm not sure I really want to do that. For what it's worth, processing the file with qpdf will create a file that is not incrementally updated and will use object streams since they are present in the original file. I hope to add support for incremental update in a future version of qpdf and will keep this issue in mind if/when I get to that.

**Comment by jberkenbilt** [comment]
I've made a note in qpdf's TODO about this issue, but I don't plan on doing anything about it in the near term. Feel free to update this issue with information you may discover from pursuing the other issues you've referenced. It might influence my decision about whether to do anything with this. There are a number of issues around qpdf offering warnings for things that it can handle but that are weird. I feel like this falls into that category. I don't know if I will ever get to those though as there are so many other things on the roadmap (pages epic, incremental updates, digital signature, and certificate-based encryption, to name just a few).

**Comment by jberkenbilt** [comment]
(Sometimes I disagree with myself.)

**Comment by calixteman** [comment]
@jberkenbilt do you have a firefox bugzilla account ? I want to share with you the pdf in the bug (the bug reporter doesn't want to make the pdf public).
When the pdf is modified  in Firefox, then when opened in Chrome there's only 1 page and in Acrobat (and pdftk) page 2 is completely empty, and as far as I can tell it's only because of the xref type mismatch. With the modified pdf, qpdf --check doesn't find any issues.

**Comment by jberkenbilt** [comment]
@calixteman Yes -- I just logged in and reactivated it. My login is ejb@ql.org.

**Comment by jberkenbilt** [comment]
@calixteman I got notified on the firefox bug and will give it a look.

**Comment by jberkenbilt** [comment]
I see the firefox bug was fixed. I'm going to close this issue, though I have a reference to it in TODO for when I get to incremental updates. Thanks.
